File: Sorting/Java/HeapSort.java
package Sorting;
// Time(all cases)		 | Space
//O(n*log(n) 			 | O(1)
public class HeapSort {
	public static void main(String[] args) {
		int arr[] = {3,37,29,5,7,4,3,2,5,6,29,7,9,3,8,21,15,16,9,55,8,10,101,57,9,12};
		System.out.println("Before heapsort the array is: ");
		printlist(arr);
		sort(arr);
		System.out.println("\nAfter heap sort the array is: ");
		printlist(arr);
	}
	
	static void printlist(int arr[])
	{
		for(int nums : arr)
		{
			System.out.print(nums+" ");
		}
	}
	
	
	
	static void sort(int arr[])
	{
//		to sort the elements
		int n = arr.length;
// 		build max heap
		for(int i = n/2 - 1 ;i>=0; i--)
		{
//			i=n/2-1 as it will give us the left and right child nodes at the first level of the heap
			heapify(arr,n,i);
		}
		
//		heap sort arranges the elements in the reverse ascending order the bigger elements occupy the last places and we get our sorted elements
		for(int j = n-1; j>=0; j--)
		{
//			swap the root node from the last element of heap and remove the root node from the heap
			
			int temp = arr[0];
			arr[0] = arr[j];
			arr[j] = temp;
			
//			this function will have size 5 instead of 6 and will remove the largest element from the heap to initiate heapsort
//			this will continue until j = 0 i.e. only one element is left and one element is sorted in itself.
//			therefore we have the sorted list
			heapify(arr,j,0);
		}
		
		
	}
	
	
	static void heapify(int [] arr, int n, int root)
	{
//		maintain the max-heap structure
//		to keep track of the original index of root  we create another index and to comapre
//		if there is any element in the heap greater than the root
		int largest = root;
//		left child node
		int left = 2 * root + 1;
//		right child node
		int right = 2 * root + 2;
		
//	check if the left child is greater than the root node
		if(left < n && arr[left] > arr[largest])
		{
			largest = left;
		}
	
//		check if right child is greater
		if(right < n && arr[right] > arr[largest])
		{
			largest = right;
		}
	
// 		compare the current root value with the largest value
// 		if root is greater then continue if not then swap that node with root
		if(largest != root)
		{
//			if the index of the largest node is not root then swap largest with the root
			int swap = arr[root];
			arr[root] = arr[largest];
			arr[largest] = swap;

//		now check if there is any element greater than the second greatest element in the array
//			if there is that means parent node is smaller so to maintain maz-heap we heapify\
			heapify(arr, n ,largest);		
			
		}
//		above condition willnot execute when we have max-heap structure maintained i.e. 
//		if the parent nodes is greater than the child node and root node is the largest i.e when largest = 0
	
	
	
	}
	
	

}

File: Sorting/Python3/Heap_Sort.py
def heapSort(array):
    # Write your code here.
    sort(array)
    return array


def sort(array):
    n = len(array)

    # build max heap
    # start from n-1//2(// for int val) to -1(excluded) decrease -1 each iteration
    for i in range(n - 1 // 2, -1, -1):
        heapify(array, n, i)

    # Heap sort
    # bigger elements are placed first in the last places in an order
    for j in range(n - 1, -1, -1):
        # swap the root element with the last node then remove it
        temp = array[0]
        array[0] = array[j]
        array[j] = temp

        # after swapping remove the root node by settting i as size
        # again the heap will be checked for max heap
        # above process of swapping and removal takes place until
        # all the elements are perfectly sorted
        heapify(array, j, 0)

# to maintain the max-heap structure
def heapify(array, n, root_idx):
    largest = root_idx
    # left child
    left_child = 2 * root_idx + 1
    # right_child
    right_child = 2 * root_idx + 2

    # check which child is larger
    if left_child < n and array[left_child] > array[largest]:
        largest = left_child
    if right_child < n and array[right_child] > array[largest]:
        largest = right_child

    # check if root node is smaller
    if largest != root_idx:
        # means there is element larger than the root
        # swap that element with the root
        swap = array[root_idx]
        array[root_idx] = array[largest]
        array[largest] = swap

        # check if the max-heap structure is maintained
        # pass largest as root node to see if there is some
        # child greater than parent
        heapify(array, n, largest)

if __name__ == '__main__':
    arr = [3,56,7,9,1,12,5]
    print('Before sorting: ',arr)
    heapSort(arr)
    print("After sorting: ",arr)
